core/arch/arm/kernel/elf_load.c

```diff
@@ -567,28 +670,38 @@ TEE_Result elf_load_body(struct elf_load_state *state, vaddr_t vabase)
 	void *p;
 	uint8_t *dst = (uint8_t *)vabase;
 	struct elf_ehdr ehdr;
-	size_t offs;
+	size_t offs = 0;
+	size_t e_p_hdr_sz;
 
 	copy_ehdr(&ehdr, state);
-
-	/*
-	 * Zero initialize everything to make sure that all memory not
-	 * updated from the ELF is zero (covering .bss and eventual gaps).
-	 */
-	memset(dst, 0, state->vasize);
+	e_p_hdr_sz = ehdr.e_phoff + ehdr.e_phnum * ehdr.e_phentsize;
 
 	/*
 	 * Copy the segments
 	 */
-	memcpy(dst, state->ta_head, state->ta_head_size);
-	offs = state->ta_head_size;
+	if (state->ta_head_size) {
+		memcpy(dst, state->ta_head, state->ta_head_size);
+		offs = state->ta_head_size;
+	}
 	for (n = 0; n < ehdr.e_phnum; n++) {
 		struct elf_phdr phdr;
 
 		copy_phdr(&phdr, state, n);
+		/*
+		 * The PT_DYNAMIC segment is always included in a PT_LOAD
+		 * segment so it can be ignored here.
+		 */
 		if (phdr.p_type != PT_LOAD)
 			continue;
 
+		if (phdr.p_offset < e_p_hdr_sz) {
+			/*
+			 * The first loadable segment contains the ELF and
```

Comment:
This is a bit cryptic to me. Please add a reference.

Reply:
I have expanded the comment. Basically we skip what we have read already because the TA store interface does not allow to read things twice.

To illustrate why it is a problem now and not before, let's examine a statically linked TA. Because of the linker script, `.ta_head` is placed at the exact beginning of the first PT_LOAD segment which starts 64K into the file for alignment reasons (note: lots of wasted space here but that's another story). So when we read the ELF, we do read the headers, then skip the padding, then reach the first loadable segment (see the `>>` mark below).
```
arm-linux-gnueabihf-readelf --program-header optee_examples/talibs/ta/93038f64-871d-4773-b1f5-878697745eee.stripped.elf 

Elf file type is DYN (Shared object file)
Entry point 0x20
There are 5 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x01c658 0x0000c658 0x0000c658 0x00860 0x00860 R   0x4
>>LOAD           0x010000 0x00000000 0x00000000 0x0a40c 0x0a40c R E 0x10000
  LOAD           0x01a40c 0x0000a40c 0x0000a40c 0x02f10 0x02f10 R   0x10000
  LOAD           0x01e000 0x0000e000 0x0000e000 0x000a4 0x0aa20 RW  0x10000
  DYNAMIC        0x01d008 0x0000d008 0x0000d008 0x000b0 0x000b0 RW  0x4

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     .ta_head .text .plt 
   02     .rodata .ARM.extab .ARM.extab.text.utee_panic .ARM.extab.text.__aeabi_ldivmod .ARM.extab.text.__aeabi_uldivmod .ARM.exidx .got .dynsym .rel.got .rel.plt .dynamic .dynstr .hash .rel.dyn 
   03     .data .bss 
   04     .dynamic 
```

With a shared library however, the first loadable segment happens to be at offset zero in the file, thus containing the ELF and program headers in addition to some sections:
```
$ arm-linux-gnueabihf-readelf --program-header optee_examples/talibs/talib3/libtalib3.stripped.so 

Elf file type is DYN (Shared object file)
Entry point 0x288
There are 5 program headers, starting at offset 52

Program Headers:
  Type           Offset   VirtAddr   PhysAddr   FileSiz MemSiz  Flg Align
  EXIDX          0x0002bc 0x000002bc 0x000002bc 0x00010 0x00010 R   0x4
>>LOAD           0x000000 0x00000000 0x00000000 0x002cc 0x002cc R E 0x10000
  LOAD           0x0002cc 0x000102cc 0x000102cc 0x00090 0x00090 RW  0x10000
  DYNAMIC        0x0002cc 0x000102cc 0x000102cc 0x00080 0x00080 RW  0x4
  GNU_STACK      0x000000 0x00000000 0x00000000 0x00000 0x00000 RW  0x10

 Section to Segment mapping:
  Segment Sections...
   00     .ARM.exidx 
   01     .hash .dynsym .dynstr .rel.plt .plt .text .rodata .ARM.exidx 
   02     .dynamic .got 
   03     .dynamic 
   04     
```

Reply:
Aha, thanks for the explanation.
Since it's the first header, wouldn't it be easier to prepare `offs` before starting the loop?

Reply:
What do you mean? Before the loop we don't know yet if the first PT_LOAD segment covers the ELF and program headers or not (there is usually EXIDX first).

Reply:
OK